src/features/debugger/debugMessage.ts

```diff
@@ -152,72 +142,86 @@ async function debugMessageDelivery(
-      const matchedEvents = (await igp.queryFilter(filter)) || [];
-      logger.debug(`Found ${matchedEvents.length} payments to IGP for msg ${messageId}`);
-      logger.debug(matchedEvents);
-      for (const payment of matchedEvents) {
-        gasAlreadyFunded = gasAlreadyFunded.add(payment.args.gasAmount);
-      }
-    } else {
+    let gasDetails: MessageDebugDetails['gasDetails'] = {
+      deliveryGasEstimate,
+    };
+    if (totalGasAmount && BigNumber.from(totalGasAmount).gt(0)) {
       logger.debug(`Using totalGasAmount info from message: ${totalGasAmount}`);
       gasAlreadyFunded = BigNumber.from(totalGasAmount);
+    } else {
+      logger.debug('Querying for gas payments events for msg to any contract');
+      const { contractToPayments, contractToTotalGas, numPayments, numIGPs } =
+        await fetchGasPaymentEvents(originProvider, messageId);
+      gasDetails = { deliveryGasEstimate, contractToPayments };
+      logger.debug(`Found ${numPayments} payments to ${numIGPs} IGPs for msg ${messageId}`);
+      if (numIGPs === 1) {
+        gasAlreadyFunded = Object.values(contractToTotalGas)[0];
+      } else if (numIGPs > 1) {
+        logger.warn(`>1 IGPs paid for msg ${messageId}. Unsure which to use, skipping check.`);
+        return { gasDetails };
+      }
     }
 
     logger.debug('Amount of gas paid for to IGP:', gasAlreadyFunded.toString());
-    logger.debug('Amount of gas required:', deliveryGasEst);
+    logger.debug('Approximate amount of gas required:', deliveryGasEstimate);
     if (gasAlreadyFunded.lte(0)) {
-      return { isFunded: false, igpDetails: 'Origin IGP has not received any gas payments' };
-    } else if (gasAlreadyFunded.lte(deliveryGasEst)) {
       return {
-        isFunded: false,
-        igpDetails: `Origin IGP gas amount is ${gasAlreadyFunded.toString()} but requires ${deliveryGasEst}`,
+        status: MessageDebugStatus.GasUnderfunded,
+        details: 'Origin IGP has not received any gas payments',
+        gasDetails,
+      };
+    } else if (gasAlreadyFunded.lte(deliveryGasEstimate)) {
+      return {
+        status: MessageDebugStatus.GasUnderfunded,
+        details: `Origin IGP gas amount is ${gasAlreadyFunded.toString()} but requires ${deliveryGasEstimate}`,
+        gasDetails,
       };
     } else {
-      return { isFunded: true, igpDetails: '' };
+      return { gasDetails };
     }
   } catch (error) {
     logger.warn('Error estimating delivery gas cost for message', error);
-    return { isFunded: true, igpDetails: '' };
+    return {};
+  }
+}
+
+async function fetchGasPaymentEvents(provider: Provider, messageId: string) {
+  const igpInterface = InterchainGasPaymaster__factory.createInterface();
+  const paymentFragment = igpInterface.getEvent('GasPayment');
+  const paymentTopics = igpInterface.encodeFilterTopics(paymentFragment.name, [messageId]);
+  const paymentLogs = (await provider.getLogs({ topics: paymentTopics })) || [];
```

Comment:
im surprised providers dont vomit about too large a block range

Reply:
They would! but SmartProvider handles this for us